ledger/src/blockstore.rs

```diff
@@ -2923,13 +2924,68 @@ impl Blockstore {
         self.last_root()
     }
 
+    pub fn lowest_cleanup_slot(&self) -> Slot {
+        *self.lowest_cleanup_slot.read().unwrap()
+    }
+
     pub fn storage_size(&self) -> Result<u64> {
         self.db.storage_size()
     }
 
     pub fn is_primary_access(&self) -> bool {
         self.db.is_primary_access()
     }
+
+    pub fn scan_and_fix_roots(&self, exit: &Arc<AtomicBool>) -> Result<()> {
+        if !self.is_root(self.last_root()) {
```

Comment:
little explanation of this special casing might help future readers... maybe is this needed for proper looping and termination of below code? And, is this observed in the wild?

Reply:
hmmm yeah seems like `last_root()` should at least be something that exists in the roots column family on startup 

```
let max_root = db
            .iter::<cf::Root>(IteratorMode::End)?
            .next()
            .map(|(slot, _)| slot)
            .unwrap_or(0);
```

and could potentially be updated by replay while this thread is running, but should always be a root, so maybe this check is redundant?

Reply:
> maybe this check is redundant?

Ah yes: https://github.com/solana-labs/solana/blob/d41266e4e915cc70fdec4a3ef54d5f1289f1afaa/ledger/src/blockstore.rs#L314-L320
This check is redundant, thank you. However, this makes me concerned that if one of these gaps is at the tip of the ledger, this scan won't catch and fix it. Thoughts?